clusterctl/clusterdeployer/clusterclient.go

```diff
@@ -0,0 +1,211 @@
+package clusterdeployer
+
+import (
+	"io/ioutil"
+	"os"
+	"os/exec"
+
+	"fmt"
+	"github.com/golang/glog"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+	"sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset"
+	"sigs.k8s.io/cluster-api/util"
+	"strings"
+	"time"
+)
+
+type clusterClient struct {
+	clientSet      clientset.Interface
+	kubeconfigFile string
+	closeFn        func() error
+}
+
+func NewClusterClient(kubeconfig string) (*clusterClient, error) {
+	f, err := ioutil.TempFile("", "")
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	f.WriteString(kubeconfig)
+	c, err := NewClusterClientFromFile(f.Name())
+	if err != nil {
+		return nil, err
+	}
+	c.closeFn = c.removeKubeconfigFile
+	return c, nil
+}
+
+func (c *clusterClient) removeKubeconfigFile() error {
+	return os.Remove(c.kubeconfigFile)
+}
+
+func NewClusterClientFromFile(kubeconfigFile string) (*clusterClient, error) {
+	c, err := util.NewClientSet(kubeconfigFile)
+	if err != nil {
+		return nil, err
+	}
+
+	return &clusterClient{
+		kubeconfigFile: kubeconfigFile,
+		clientSet:      c,
+	}, nil
+}
+
+// Frees resources associated with the cluster client
+func (c *clusterClient) Close() error {
+	if c.closeFn != nil {
+		return c.closeFn()
+	}
+	return nil
+}
+
+func (c *clusterClient) Apply(manifest string) error {
+	return c.waitForKubectlApply(manifest)
+}
+
+func (c *clusterClient) GetClusterObjects() ([]*clusterv1.Cluster, error) {
+	clusters := []*clusterv1.Cluster{}
+	// TODO: Iterate over all namespaces where we could have Cluster API Objects
+	clusterlist, err := c.clientSet.ClusterV1alpha1().Clusters(apiv1.NamespaceDefault).List(metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, cluster := range clusterlist.Items {
+		cluster, err := c.clientSet.ClusterV1alpha1().Clusters(apiv1.NamespaceDefault).Get(cluster.Name, metav1.GetOptions{})
```

Comment:
instead of apiv1.NamespaceDefault, better to use the namespace on the cluster object? cluster.Namespace?

Reply:
Fixed.